### ISSUE TITLE
[Youtube] -S codec:h264:aac selects 22 format

### ISSUE THREAD

**Issue by MrRawes** [yt-dlp/yt-dlp#3648]
### Checklist

- [X] I'm reporting a bug unrelated to a specific site
- [X] I've verified that I'm running yt-dlp version **2022.04.08** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later, [commit ff4d7860d50407f8a1daa1094f65300e8455ec92](https://github.com/yt-dlp/yt-dlp/commit/ff4d7860d50407f8a1daa1094f65300e8455ec92)
- [X] I've checked that all provided URLs are alive and playable in a browser
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/ytdl-org/youtube-dl#video-url-contains-an-ampersand-and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues including closed ones. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)

### Description

when using `-S codec:h264:aac` yt-dlp selects 22 on youtube instead of 137+140

### Verbose log

```shell
[debug] Command-line config: ['--no-config', '-sv', '-S', 'codec:h264:aac', 'https://www.youtube.com/watch?v=BaW_jenozKc']
[debug] Encodings: locale UTF-8, fs utf-8, out utf-8 (No ANSI), err utf-8 (No ANSI), pref UTF-8
[debug] yt-dlp version 2022.04.08 [7884ade65]
[debug] Lazy loading extractors is disabled
[debug] Python version 3.10.4 (CPython 64bit) - Linux-5.17.0-1-amd64-x86_64-with-glibc2.33
[debug] Checking exe version: ffprobe -bsfs
[debug] Checking exe version: ffmpeg -bsfs
[debug] exe versions: ffmpeg 4.4.2-1 (setts), ffprobe 4.4.2-1
[debug] Optional libraries: Cryptodome-3.14.1, brotli-1.0.9, certifi-2021.10.08, mutagen-1.45.1, sqlite3-2.6.0, websockets-10.3
[debug] Proxy map: {}
[debug] [youtube] Extracting URL: https://www.youtube.com/watch?v=BaW_jenozKc
[youtube] BaW_jenozKc: Downloading webpage
[youtube] BaW_jenozKc: Downloading android player API JSON
[debug] Sort order given by user: codec:h264:aac
[debug] Sort order given by extractor: quality, res, fps, hdr:12, source, codec:vp9.2, lang, proto
[debug] Formats sorted by: hasvid, ie_pref, vcodec:h264(7), acodec:aac(9), quality, res, fps, hdr:12(7), source, lang, proto, filesize, fs_approx, tbr, vbr, abr, asr, vext, aext, hasaud, id
[debug] Default format spec: bestvideo*+bestaudio/best
[info] BaW_jenozKc: Downloading 1 format(s): 22
```

**Comment by pukkandan** [comment]
Looks correct to me. Why do you say it should select 137+140?

**Comment by MrRawes** [comment]
> Looks correct to me. Why do you say it should select 137+140?

because `-S vcodec:h264` selects 137 and `-S acodec:aac` selects 140, but `-S codec:h264:aac` or `vcodec:h264,acodec:aac` selects a format that is of lesser quality in both video and audio (22 in this case)

**Comment by pukkandan** [comment]
> -S acodec:aac selects 140

No

```
❯ yt-dlp -S acodec:aac -O format BaW_jenozKc
22 - 1280x720 (720p)
```

When using `-S codec:h264:aac`, you are asking yt-dlp to prioritize h264+acc above anything else. So format 22 which satisfies both gets the highest priority

```
❯ yt-dlp -S codec:h264:aac -F BaW_jenozKc
[youtube] BaW_jenozKc: Downloading webpage
[youtube] BaW_jenozKc: Downloading android player API JSON
[info] Available formats for BaW_jenozKc:
ID  EXT  RESOLUTION FPS │   FILESIZE   TBR PROTO │ VCODEC        VBR ACODEC      ABR     ASR MORE INFO
──────────────────────────────────────────────────────────────────────────────────────────────────────────────
249 webm audio only     │   58.17KiB   48k https │ audio only        opus        48k 48000Hz low, webm_dash
250 webm audio only     │   76.07KiB   63k https │ audio only        opus        63k 48000Hz low, webm_dash
251 webm audio only     │  138.96KiB  115k https │ audio only        opus       115k 48000Hz medium, webm_dash
139 m4a  audio only     │   58.59KiB   48k https │ audio only        mp4a.40.5   48k 22050Hz low, m4a_dash
140 m4a  audio only     │  154.06KiB  127k https │ audio only        mp4a.40.2  127k 44100Hz medium, m4a_dash
17  3gp  176x144     12 │   55.79KiB   44k https │ mp4v.20.3     44k mp4a.40.2    0k 22050Hz 144p
278 webm 256x144     30 │   52.22KiB   43k https │ vp9           43k video only              144p, webm_dash
242 webm 426x240     30 │   33.27KiB   27k https │ vp9           27k video only              240p, webm_dash
243 webm 640x360     30 │   75.55KiB   62k https │ vp9           62k video only              360p, webm_dash
244 webm 854x480     30 │  165.49KiB  137k https │ vp9          137k video only              480p, webm_dash
247 webm 1280x720    30 │  504.68KiB  420k https │ vp9          420k video only              720p, webm_dash
248 webm 1920x1080   30 │  965.31KiB  804k https │ vp9          804k video only              1080p, webm_dash
160 mp4  256x144     15 │  135.08KiB  112k https │ avc1.4d400c  112k video only              144p, mp4_dash
133 mp4  426x240     30 │  294.27KiB  245k https │ avc1.4d4015  245k video only              240p, mp4_dash
134 mp4  640x360     30 │  349.59KiB  292k https │ avc1.4d401e  292k video only              360p, mp4_dash
135 mp4  854x480     30 │  849.41KiB  710k https │ avc1.4d401f  710k video only              480p, mp4_dash
136 mp4  1280x720    30 │    1.60MiB 1365k https │ avc1.4d401f 1365k video only              720p, mp4_dash
137 mp4  1920x1080   30 │    2.11MiB 1803k https │ avc1.640028 1803k video only              1080p, mp4_dash
18  mp4  640x360     30 │  354.29KiB  293k https │ avc1.42001E  293k mp4a.40.2    0k 44100Hz 360p
22  mp4  1280x720    30 │ ~  1.82MiB 1493k https │ avc1.64001F 1493k mp4a.40.2    0k 44100Hz 720p
```

I think what you are looking for is `-S res,codec:h264:aac` or maybe `-S vcodec:h264,res,acodec:aac`

**Comment by pukkandan** [comment]
> and when using -f "(bv*[vcodec^=avc1]/bv*)+(ba[acodec^=mp4a]/ba)" it selects 137+140

There is often no one-to-one translation b/w `-S` and `-f`, but the closest `-f` to `-S codec:h264:aac` is:
```
❯ yt-dlp BaW_jenozKc -f "(bv*[vcodec^=avc1][acodec^=mp4a]/bv*[vcodec^=avc1]/bv*)+(ba[acodec^=mp4a]/ba)" -O format
22 - 1280x720 (720p)
```

**Comment by MikeChristianson** [comment]
In case someone else ends up here when trying to find out why format `22` is unexpectedly selected…

It turns out there can be a discrepancy between what `-O format` yields and what is actually downloaded, depending on the presence of `ffmpeg`. 

Here's an example using `-S "vcodec:h264,res,acodec:aac"`:

## TLDR Download output difference
```diff
6c6
< [debug] exe versions: none
---
> [debug] exe versions: ffmpeg 4.4.2 (setts), ffprobe 4.4.2
14c14
< [debug] [youtube] Decrypted nsig 55XMPTn9fCDHOkX => RlvOMB28Nxv_WQ
---
> [debug] [youtube] Decrypted nsig dxZpS1wUM6TO5nS => lKCBegQgqTiCcg
19,20c19,20
< [debug] Default format spec: best/bestvideo+bestaudio
< [info] nU2cw-ta_l0: Downloading 1 format(s): 22
---
> [debug] Default format spec: bestvideo*+bestaudio/best
> [info] nU2cw-ta_l0: Downloading 1 format(s): 299+140
```

## `-O format`
```
❯ yt-dlp -S "vcodec:h264,res,acodec:aac" -O format 'https://www.youtube.com/watch?v=nU2cw-ta_l0'                                                         
299 - 1920x1080 (1080p60)+140 - audio only (medium)
```

## Download without `ffmpeg`
```
❯ yt-dlp -v -S "vcodec:h264,res,acodec:aac" 'https://www.youtube.com/watch?v=nU2cw-ta_l0'
[debug] Command-line config: ['-v', '-S', 'vcodec:h264,res,acodec:aac', 'https://www.youtube.com/watch?v=nU2cw-ta_l0']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version stable@2023.03.04 [392389b7d] (zip)
[debug] Python 3.10.6 (CPython x86_64 64bit) - Linux-5.15.0-69-generic-x86_64-with-glibc2.35 (OpenSSL 3.0.2 15 Mar 2022, glibc 2.35)
[debug] exe versions: none
[debug] Optional libraries: certifi-2020.06.20, no_Cryptodome-None, secretstorage-3.3.1, sqlite3-2.6.0
[debug] Proxy map: {}
[debug] Loaded 1786 extractors
[youtube] Extracting URL: https://www.youtube.com/watch?v=nU2cw-ta_l0
[youtube] nU2cw-ta_l0: Downloading webpage
[youtube] nU2cw-ta_l0: Downloading android player API JSON
[debug] Loading youtube-nsig.f55759b8 from cache
[debug] [youtube] Decrypted nsig 55XMPTn9fCDHOkX => RlvOMB28Nxv_WQ
[debug] Sort order given by user: vcodec:h264, res, acodec:aac
[debug] Sort order given by extractor: quality, res, fps, hdr:12, source, vcodec:vp9.2, channels, acodec, lang, proto
[debug] Formats sorted by: hasvid, ie_pref, vcodec:h264(7), res, acodec:aac(10), quality, fps, hdr:12(7), source, channels, lang, proto, filesize, fs_appr
ox, tbr, vbr, abr, asr, vext, aext, hasaud, id
[debug] Default format spec: best/bestvideo+bestaudio
[info] nU2cw-ta_l0: Downloading 1 format(s): 22
```

## Download with `ffmpeg`
```
❯ yt-dlp -v -S "vcodec:h264,res,acodec:aac" 'https://www.youtube.com/watch?v=nU2cw-ta_l0'
[debug] Command-line config: ['-v', '-S', 'vcodec:h264,res,acodec:aac', 'https://www.youtube.com/watch?v=nU2cw-ta_l0']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version stable@2023.03.04 [392389b7d] (zip)
[debug] Python 3.10.6 (CPython x86_64 64bit) - Linux-5.15.0-69-generic-x86_64-with-glibc2.35 (OpenSSL 3.0.2 15 Mar 2022, glibc 2.35)
[debug] exe versions: ffmpeg 4.4.2 (setts), ffprobe 4.4.2
[debug] Optional libraries: certifi-2020.06.20, no_Cryptodome-None, secretstorage-3.3.1, sqlite3-2.6.0
[debug] Proxy map: {}
[debug] Loaded 1786 extractors
[youtube] Extracting URL: https://www.youtube.com/watch?v=nU2cw-ta_l0
[youtube] nU2cw-ta_l0: Downloading webpage
[youtube] nU2cw-ta_l0: Downloading android player API JSON
[debug] Loading youtube-nsig.f55759b8 from cache
[debug] [youtube] Decrypted nsig dxZpS1wUM6TO5nS => lKCBegQgqTiCcg
[debug] Sort order given by user: vcodec:h264, res, acodec:aac
[debug] Sort order given by extractor: quality, res, fps, hdr:12, source, vcodec:vp9.2, channels, acodec, lang, proto
[debug] Formats sorted by: hasvid, ie_pref, vcodec:h264(7), res, acodec:aac(10), quality, fps, hdr:12(7), source, channels, lang, proto, filesize, fs_appr
ox, tbr, vbr, abr, asr, vext, aext, hasaud, id
[debug] Default format spec: bestvideo*+bestaudio/best
[info] nU2cw-ta_l0: Downloading 1 format(s): 299+140
```

**Comment by pukkandan** [comment]
-O format without ffmpeg should match what is downloaded without ffmpeg.
